test/utils_test.js

```diff
@@ -242,6 +242,207 @@ describe("utils", () => {
+        foo: 42,
+      });
+    });
+
+    it("should work for nested array", () => {
+      const schema = {
+        type: "object",
+        properties: {
+          arr: {
+            type: "array",
+            items: {
+              type: "boolean",
+              default: false,
+            },
+          },
+        },
+      };
+
+      expect(
+        getDefaultFormState(schema, {
+          arr: [undefined],
+        })
+      ).eql({
+        arr: [false],
+      });
+    });
+
+    it("should work for arrays of objects with defaults", () => {
+      const schema = {
+        type: "object",
+        properties: {
+          arr: {
+            type: "array",
+            items: {
+              type: "object",
+              properties: {
+                defaultFilled: {
+                  type: "boolean",
+                  default: true,
+                },
+              },
+            },
+          },
+        },
+      };
+
+      expect(
+        getDefaultFormState(schema, {
+          arr: [{}],
+        })
+      ).eql({
+        arr: [{ defaultFilled: true }],
+      });
+    });
+
+    it("should use additionalItems defaults for items not specified in items", () => {
+      const schema = {
+        type: "object",
+        properties: {
+          arr: {
+            type: "array",
+            items: [
+              {
+                type: "boolean",
+                default: false,
+              },
+            ],
+            additionalItems: {
+              type: "string",
+              default: "foobar",
+            },
+          },
+        },
+      };
+
+      expect(
+        getDefaultFormState(schema, {
+          arr: [undefined, undefined],
+        })
+      ).eql({
+        arr: [false, "foobar"],
+      });
+    });
+
+    it("should use array default in favor of array item default", () => {
+      const schema = {
+        type: "array",
+        default: [1, 2],
+        items: {
+          type: "number",
+          default: 3,
+        },
+      };
+
+      expect(getDefaultFormState(schema, [0, undefined])).eql([0, 2]);
+    });
+
+    it("should fill defaults to an array with less items than minItems", () => {
+      const schema = {
+        type: "array",
+        minItems: 2,
+        items: {
+          type: "number",
+          default: 0,
+        },
+      };
+
+      expect(getDefaultFormState(schema, [1])).eql([1, 0]);
+    });
+
+    it("should use array  array default in favor or array item default even if it is a ref", () => {
+      const schema = {
+        type: "array",
+        default: [0],
+        items: {
+          $ref: "#/definitions/def",
+        },
+        definitions: {
+          def: {
+            type: "number",
+            default: 1,
+          },
+        },
+      };
+
+      expect(getDefaultFormState(schema, [undefined], schema.definitions)).eql([
+        0,
+      ]);
+    });
```

Comment:
Same concern about filling in the toplevel default when the element exists. I think this test would make sense to me if instead of `[undefined]`, you had `undefined`. However, as written, I would have expected it to be `[1]`.